w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+"""
+hpp.py
+ 
+Copyright 2006 Andres Riancho
+ 
+This file is part of w3af, http://w3af.org/ .
+ 
+w3af is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation version 2 of the License.
+ 
+w3af is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+ 
+You should have received a copy of the GNU General Public License
+along with w3af; if not, write to the Free Software
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+"""
+
+from io import StringIO
+from w3af.core.controllers.plugins.audit_plugin import AuditPlugin
+from w3af.core.data.fuzzer.fuzzer import create_mutants
+from w3af.core.data.kb.vuln import Vuln
+from w3af.core.data.kb.info import Info
+from w3af.core.data.fuzzer.mutants.querystring_mutant import QSMutant
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.constants.severity as severity
+from w3af.core.controllers.misc.fuzzy_string_cmp import fuzzy_equal
+import w3af.core.data.parsers.parser_cache as parser_cache
+from lxml import etree
+from w3af.core.controllers.exceptions import BaseFrameworkException
+from w3af.core.data.parsers.url import URL
+from w3af.core.data.kb.info_set import InfoSet
+
+
+REFLECT_TOKEN = 'w3afHPPReflectToken'
+FORGE_TOKEN = 'w3afHPPForgeToken'
+EQUALITY_THRESHOLD = 0.95
+#paramater precedence types
+LAST = 'last occurance'
+FIRST = 'first occurance'
+CONCAT = 'concatination'
+UNK = 'unknown'
+
+
+class hpp(AuditPlugin):
+    """
+    Identify http parameter pollution vulnerabilities.
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+        self.precedence = None
+        self.rlink = ''
+
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+
+        :param freq: A FuzzableRequest
+
+        """
+
+        rparams = {}
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+        method = freq.get_method()
+        self._orig_uri = freq.get_uri()
```

Comment:
`self._orig_uri = freq.get_uri()` won't work, since we run audit plugins in different threads and thread #1 will overwrite the `self._orig_uri = freq.get_uri()` set by thread #2 . You should use a local variable